商品削除機能

AiAoHaru opened the pull request:
# What
- 出品した商品の削除機能の追加（出品者だけが商品情報を削除可能）

# Why
- 出品者以外が不正に商品を削除できない様にするため

AiAoHaru committed:
コントローラーのdestroyアクションの設定

AiAoHaru commented:
## 挙動確認の前提条件
- ユーザー1（test太郎）が商品を出品

#### ①一覧表示機能（トップ画面）が未実装の為、直接詳細リンクへの遷移で挙動を確認


## ※挙動確認のリンク
- ユーザー1（test太郎）ログイン状態から詳細ページへの遷移し出品済み商品の削除を実行
https://gyazo.com/f3feeba9f533dac1307ea8dc5f73403f
商品削除前SequelPro：https://gyazo.com/eed7b47834cf0f9ee6424d6dc554e282
商品削除後SequelPro：https://gyazo.com/bbd98feee98f48c7e28a621b41c9cd07

- ユーザー2（test花子）がユーザー1（test太郎）の出品商品詳細を閲覧はできるが削除できないことを確認（削除ボタンの有無確認）
https://gyazo.com/f9f09b2b9c4bfe302a3b4a5c1f9ce680
- 未ログインユーザーがユーザー1（test太郎）の出品商品詳細を閲覧はできるが削除できないことを確認（削除ボタンの有無確認）
https://gyazo.com/7d2f50304f6ab5630d7570ab0c4e08d8

AiAoHaru committed:
追加アレンジ実装（販売済み出品商品の編集不可設定）

販売（購入）済みの商品に関しては、出品者の編集ボタンを非表示に変更

※見本では、出品時に設定した項目の全てが販売済み商品でも変更可能となっている。

◉今後の検討について
※削除機能については、現状エラーとなる。
→購入済み商品についての削除ボタンは今後検討が必要。
　　①削除できない様にボタンを非表示にする
　　②削除しようとすると詳細ページに戻し、エラーメッセージを表示させる　等

AiAoHaru commented:
## 追加アレンジ実装
- 販売（購入）済み商品が編集可能となってしまっている
→販売（購入）済みの商品に関しては、出品者の**編集**ボタンを**非表示に変更**

#### 購入機能が未実装のため、購入処理に関する履歴をSequel Proで手動にて設定し確認する

## ※挙動確認のリンク
https://gyazo.com/d13961497b75f6c50bef70ee62815197


#### 現状について

※見本では、出品時に設定した項目の全てが販売済み商品でも変更可能となっている。

※削除機能については、現状エラーとなる。
→購入済み商品についての削除ボタンは今後検討が必要。
　　①削除できない様にボタンを非表示にする
　　②削除しようとすると詳細ページに戻し、エラーメッセージを表示させる　等

Akinori0123 commented on app/views/items/show.html.erb:
```diff
@@ -27,8 +27,10 @@
 
     <%# ログインしているユーザと出品しているユーザが同一人物である時、商品の編集と削除を表示にしましょう%>
     <% if user_signed_in? && current_user.id == @item.user_id %>    <%# ログイン済のユーザーかつ出品者の場合、商品編集と削除ボタンを表示させる条件分岐文 %>
+      <%if Purchaser.find_by(item_id: @item) == nil %>
```
30行目は「購入済みの商品は編集ボタンを表示しない」といった条件分岐でよろしいでしょうか？
編集ボタンの表示に関しては必須機能には入っていませんが、確かにアプリとして購入済みの商品の情報を編集できるのは適切ではないためとても良い考え方ですね。
現在の記述のままでも問題ありませんが、ItemとPurchaserはアソシエーションで繋がっているので「@itemに紐づいているpurchaserがあるかどうか」の方が記述としてはシンプルかと思われます。Purchaser.find_byを使わず、@item.◯◯を使う方針もトライしてみましょう。
（こちらはこのプルリクエスト上では修正必須ではありません）

AiAoHaru replied:
レビューありがとうございます。
また、アドバイスありがとうございます。
トライしてみました。

アソシエーション関係がある場合にデータが存在するかどうかという確認方法の観点を新しく習得できました。
ありがとうございます。

ちなみに、このアソシエーション関係では、belongs_toやhas_manyではデータがどれか判断がつかないため、has_oneのアソシエーション関係のみでしか成り立たないという認識でよろしいでしょうか？

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -18,6 +18,9 @@ def create
   
     def destroy
+      item = Item.find(params[:id])
+      item.destroy
```
22行目
現在商品の削除が失敗した時も次の行が実行されてしまいます。
保守性を高めるためにもし商品削除に失敗した場合は前の画面を表示するなどのエラーハンドリングをif文で実装しましょう。createアクションの記述を参考にしていただければと思います。
※エラーハンドリングとは、プログラムの処理中正常に動かなかった場合に備え、その処理をエラーとして対処する処理のことです。

AiAoHaru replied:
レビューありがとうございます。
修正しました。

Akinori0123 reviewed (commented):
レビュー依頼ありがとうございます。
コメントしましたので、ご確認をお願いいたします。

AiAoHaru committed:
削除の際のエラーハンドリングの実装

保守性を高めるために、商品削除に失敗した場合は前の画面を表示するエラーハンドリングを実装

※エラーハンドリングとは、プログラムの処理中正常に動かなかった場合に備え、その処理をエラーとして対処する処理のこと

AiAoHaru committed:
インデントの修正

AiAoHaru committed:
追加アレンジ実装（dependentオプションの設定）

◉現状出品者の詳細画面には、購入済み商品の削除ボタンが表示されている。
※対応策として
→①削除ボタンを非表示にして、販売済みの商品は削除不可にする。
→②削除できる様にする。その際、購入者情報も削除される様に実装する。

◉今回は、 購入済の商品を出品者が削除した場合に、購入者情報（parchaserのレコード）も一緒に削除させる設定とする。外部キー制約のエラー対応として。

※（今後、購入者情報のみは残せる仕様など検討）
登録ユーザーの削除（退会）機能などを追加実装した際には、userモデルに記載のアソシエーションについても、dependent::destroyオプションの設定を検討する必要がある。
ただし、購入者情報などが消えて、商品が残ってしまうと再度購入ができてしまうなどの不具合が生じるため、慎重に検討していく必要あり。
例えば、仮で退会済みユーザーなどを作っておいて、退会した場合は、その退会済みユーザーに紐つけるなど。

AiAoHaru committed:
追加アレンジ実装のファクタリング（条件分岐式）

◉Purchaser.find_byを使わず、アソシエーション関係のあることを考え、「@itemに紐づいているpurchaserがあるかどうか」の方が記述としてはシンプルになるため。

ただし、アソシエーション関係でも、belongs_toやhas_manyでは成り立たない。

AiAoHaru commented:
### 追加アレンジ実装（dependentオプションの設定）について

◉現状出品者の詳細画面には、購入済み商品の削除ボタンが表示されている。
削除実行をすると、purchasersとアソシエーション関係があるため、エラーが生じる。

◉今回は、 購入済の商品を出品者が削除した場合に、購入者情報（parchaserのレコード）も一緒に削除させる設定とする。外部キー制約のエラー対応として。

※検討課題（今後、購入者情報のみは残せる仕様など検討）
登録ユーザーの削除（退会）機能などを追加実装した際には、userモデルに記載のアソシエーションについても、dependent::destroyオプションの設定を検討する必要がある。ただし、購入者情報などが消えて、商品が残ってしまうと再度購入ができてしまうなどの不具合が生じるため、慎重に検討していく必要あり。
例えば、仮で退会済みユーザーなどを作っておいて、退会した場合は、その退会済みユーザーに紐つけるなど。

shingoteshima commented on app/controllers/items_controller.rb:
```diff
@@ -17,18 +17,24 @@ def create
-    def show
-      @item = Item.find(params[:id])
+  def destroy
+    item = Item.find(params[:id])
```
set_itemメソッドをprivate以下に作成し、before_actionで呼びましょう。
他の箇所も同様です。

AiAoHaru replied:
レビューありがとうございます。
修正しました。

shingoteshima reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認お願いいたします！

AiAoHaru committed:
set_itemメソッド及びbefore_actionの実装

set_itemメソッドをprivate以下に作成し、before_actionでの呼び出しに修正。
